torch/_torch_docs.py

```diff
@@ -740,6 +740,10 @@ def merge_dicts(*dicts):
     :meth:`torch.Tensor.expand`, are easier to read and are therefore more
     advisable to use.
 
+.. warning::
```

Comment:
Hey @ysiraichi, thanks for the ping and sorry to make you wait for a review. Now let's talk about `as_strided` and `empty_strided` and use this opportunity to correct and modernize their documentation.

First, I don't think the documentation for this function is correct, and that's part of what has made this PR hard to review. Let's look at an example program:

```
t = torch.tensor((1, 2, 3, 4, 5, 6, 7, 8, 9, 10))[::2]
# t = tensor([1, 3, 5, 7, 9])
v = torch.as_strided(t, size=[2], stride=[1])
# v = tensor([1, 2])
```

In what sense is `v` a view of `t`? It contains a value not found in `t`!

And when we look at the implementation of `as_strided` this makes sense:

https://github.com/pytorch/pytorch/blob/0a3cf8886a68e9a965f669823e640cd95ca2692e/aten/src/ATen/native/TensorShape.cpp#L769

because it seems like `as_strided` creates a view of :attr:\`input\`'s storage, not a view of :attr:\`input\`. So first we need to explain that.

Now let's talk about warning. It's trying to tell us two things:

- creating an "overlapped" tensor where multiple indices refer to the same element is a bad idea
- you probably shouldn't use this function

This seems like a strange warning to me, so let's dig into its history: https://github.com/pytorch/pytorch/pull/22842#issuecomment-511896248. Seems like there were concerns about exposing `as_strided` at all, which is reasonable, but I think we can do a better job with this warning.

What would you think of changing the warning to this, instead:

"Prefer using other view functions, like :meth:\`torch.Tensor.expand\`, to setting a view's strides manually with \`as_strided\`, as this function's behavior depends on the implementation of a tensor's storage. The constructed view of the storage must only refer to elements within the storage or a runtime error will be thrown, and if the view is "overlapped" (with multiple indices referring to the same element in memory) its behavior is undefined."

This more clearly communicates why this function shouldn't be used -- because storage is an implementation detail and relying on it is inherently risky and strange for users. This also simplifies the other warnings.

Reply:
That sounds way better! Thanks, @mruberry. I will change it right away!